// ngx-ace-wrapper/dist/lib/ace.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export var ACE_CONFIG = new InjectionToken('ACE_CONFIG');
/** @type {?} */
export var AceEditorEvents = [
    'blur',
    'focus',
    'copy',
    'paste',
    'change',
    'changeSession'
];
/** @type {?} */
export var AceSelectionEvents = [
    'changeCursor',
    'changeSelection'
];
/**
 * @record
 */
export function AceConfigInterface() { }
if (false) {
    /** @type {?|undefined} */
    AceConfigInterface.prototype.readOnly;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.cursorStyle;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.selectionStyle;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.mergeUndoDeltas;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.behavioursEnabled;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.highlightActiveLine;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.highlightSelectedWord;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.wrapBehavioursEnabled;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.copyWithEmptySelection;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.navigateWithinSoftTabs;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.autoScrollEditorIntoView;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.mode;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.wrap;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.tabSize;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.overwrite;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.useWorker;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.foldStyle;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.newLineMode;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.useSoftTabs;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.firstLineNumber;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.theme;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.minLines;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.maxLines;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.fontSize;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.fontFamily;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.showGutter;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.printMargin;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.scrollPastEnd;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.animatedScroll;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.showInvisibles;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.fadeFoldWidgets;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.showFoldWidgets;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.showLineNumbers;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.showPrintMargin;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.fixedWidthGutter;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.printMarginColumn;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.displayIndentGuides;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.highlightGutterLine;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.hScrollBarAlwaysVisible;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.vScrollBarAlwaysVisible;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.dragDelay;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.dragEnabled;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.scrollSpeed;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.focusTimeout;
    /** @type {?|undefined} */
    AceConfigInterface.prototype.tooltipFollowsMouse;
}
var AceConfig = /** @class */ (function () {
    function AceConfig(config) {
        if (config === void 0) { config = {}; }
        this.assign(config);
    }
    /**
     * @param {?=} config
     * @param {?=} target
     * @return {?}
     */
    AceConfig.prototype.assign = /**
     * @param {?=} config
     * @param {?=} target
     * @return {?}
     */
    function (config, target) {
        if (config === void 0) { config = {}; }
        target = target || this;
        for (var key in config) {
            if (config[key] && !Array.isArray(config[key]) && typeof config[key] === 'object') {
                target[key] = {};
                this.assign(config[key], target[key]);
            }
            else {
                target[key] = config[key];
            }
        }
    };
    return AceConfig;
}());
export { AceConfig };
if (false) {
    /** @type {?} */
    AceConfig.prototype.readOnly;
    /** @type {?} */
    AceConfig.prototype.cursorStyle;
    /** @type {?} */
    AceConfig.prototype.selectionStyle;
    /** @type {?} */
    AceConfig.prototype.mergeUndoDeltas;
    /** @type {?} */
    AceConfig.prototype.behavioursEnabled;
    /** @type {?} */
    AceConfig.prototype.highlightActiveLine;
    /** @type {?} */
    AceConfig.prototype.highlightSelectedWord;
    /** @type {?} */
    AceConfig.prototype.wrapBehavioursEnabled;
    /** @type {?} */
    AceConfig.prototype.copyWithEmptySelection;
    /** @type {?} */
    AceConfig.prototype.navigateWithinSoftTabs;
    /** @type {?} */
    AceConfig.prototype.autoScrollEditorIntoView;
    /** @type {?} */
    AceConfig.prototype.mode;
    /** @type {?} */
    AceConfig.prototype.wrap;
    /** @type {?} */
    AceConfig.prototype.tabSize;
    /** @type {?} */
    AceConfig.prototype.overwrite;
    /** @type {?} */
    AceConfig.prototype.useWorker;
    /** @type {?} */
    AceConfig.prototype.foldStyle;
    /** @type {?} */
    AceConfig.prototype.newLineMode;
    /** @type {?} */
    AceConfig.prototype.useSoftTabs;
    /** @type {?} */
    AceConfig.prototype.firstLineNumber;
    /** @type {?} */
    AceConfig.prototype.theme;
    /** @type {?} */
    AceConfig.prototype.minLines;
    /** @type {?} */
    AceConfig.prototype.maxLines;
    /** @type {?} */
    AceConfig.prototype.fontSize;
    /** @type {?} */
    AceConfig.prototype.fontFamily;
    /** @type {?} */
    AceConfig.prototype.showGutter;
    /** @type {?} */
    AceConfig.prototype.printMargin;
    /** @type {?} */
    AceConfig.prototype.scrollPastEnd;
    /** @type {?} */
    AceConfig.prototype.animatedScroll;
    /** @type {?} */
    AceConfig.prototype.showInvisibles;
    /** @type {?} */
    AceConfig.prototype.fadeFoldWidgets;
    /** @type {?} */
    AceConfig.prototype.showLineNumbers;
    /** @type {?} */
    AceConfig.prototype.showFoldWidgets;
    /** @type {?} */
    AceConfig.prototype.showPrintMargin;
    /** @type {?} */
    AceConfig.prototype.fixedWidthGutter;
    /** @type {?} */
    AceConfig.prototype.printMarginColumn;
    /** @type {?} */
    AceConfig.prototype.displayIndentGuides;
    /** @type {?} */
    AceConfig.prototype.highlightGutterLine;
    /** @type {?} */
    AceConfig.prototype.hScrollBarAlwaysVisible;
    /** @type {?} */
    AceConfig.prototype.vScrollBarAlwaysVisible;
    /** @type {?} */
    AceConfig.prototype.dragDelay;
    /** @type {?} */
    AceConfig.prototype.dragEnabled;
    /** @type {?} */
    AceConfig.prototype.scrollSpeed;
    /** @type {?} */
    AceConfig.prototype.focusTimeout;
    /** @type {?} */
    AceConfig.prototype.tooltipFollowsMouse;
}
