// @ngstack/translate/fesm2015/ngstack-translate.js
import { __awaiter } from 'tslib';
import { Injectable, EventEmitter, Inject, InjectionToken, Optional, NgModule, Pipe, defineInjectable, inject } from '@angular/core';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { Title } from '@angular/platform-browser';
import { Subject } from 'rxjs';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const TRANSLATE_SETTINGS = new InjectionToken('TRANSLATE_SETTINGS');
class TranslateService {
    /**
     * @param {?} http
     * @param {?} settings
     */
    constructor(http, settings) {
        this.http = http;
        this.data = {};
        this._fallbackLang = 'en';
        this._translationRoot = 'assets/i18n';
        /**
         * Raised each time active language gets changed.
         */
        this.activeLangChanged = new EventEmitter();
        /**
         * Toggles debug mode.
         *
         * When in the debug mode, the service automatically prepends active language id to very translated result.
         * That allows to verify that your components support i18n correctly and do not contain hard-coded text.
         */
        this.debugMode = false;
        /**
         * Disable caching and always download language files.
         *
         * Applies cache busting query parameters to urls, for example: '?v=1522426955882'.
         */
        this.disableCache = false;
        /**
         * List of supported languages.
         *
         * The service will attempt to load resource files only for given set of languages,
         * and will automatically use fallback language for all unspecified values.
         *
         * By default this property is empty and service is going to probe all language files.
         * Active and Fallback languages are always taken into account even if you do not specify them in the list.
         */
        this.supportedLangs = [];
        /**
         * List of extra paths to look for translation files.
         *
         * By default this property is empty.
         * The value of `translationRoot` property is always taken into account.
         */
        this.translatePaths = [];
        this.applySettings(settings);
    }
    /**
     * The fallback language to use when a resource string for the active language is not available.
     * @return {?}
     */
    get fallbackLang() {
        return this._fallbackLang;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set fallbackLang(value) {
        this._fallbackLang = value || 'en';
    }
    /**
     * The language to use for the translations.
     * @return {?}
     */
    get activeLang() {
        return this._activeLang;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set activeLang(value) {
        /** @type {?} */
        const previousValue = this._activeLang;
        /** @type {?} */
        const newValue = value || this.fallbackLang;
        /** @type {?} */
        const changed = newValue !== previousValue;
        if (changed) {
            this._activeLang = newValue;
            this.use(newValue).then(() => {
                this.activeLangChanged.next({
                    previousValue: previousValue,
                    currentValue: newValue
                });
            });
        }
    }
    /**
     * @return {?}
     */
    getBrowserLanguage() {
        const [lang /*, locale*/] = navigator.language
            .replace('-', '_')
            .toLowerCase()
            .split('_');
        return lang;
    }
    /**
     * The root path to use when loading default translation files.
     * Defaults to 'assets/i18n'.
     * @return {?}
     */
    get translationRoot() {
        return this._translationRoot;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set translationRoot(value) {
        this._translationRoot = value || 'assets/i18n';
    }
    /**
     * @param {?} settings
     * @return {?}
     */
    applySettings(settings) {
        const { debugMode, disableCache, supportedLangs, translatePaths, translationRoot, activeLang } = this;
        /** @type {?} */
        const defaults = Object.assign({ debugMode,
            disableCache,
            supportedLangs,
            translatePaths,
            translationRoot,
            activeLang }, settings);
        this.debugMode = defaults.debugMode;
        this.disableCache = defaults.disableCache;
        this.supportedLangs = defaults.supportedLangs;
        this.translatePaths = defaults.translatePaths;
        this.translationRoot = defaults.translationRoot;
        this._activeLang = defaults.activeLang || this.getBrowserLanguage();
    }
    /**
     * Get translated string
     *
     * \@memberof TranslateService
     * @param {?} key Translation key
     * @param {?=} params
     * @param {?=} lang
     * @return {?} Translated string
     */
    get(key, params, lang) {
        if (key) {
            /** @type {?} */
            let value = this.getValue(lang || this.activeLang, key);
            if (value === key) {
                value = this.getValue(this.fallbackLang, key);
            }
            return this.format(value, params);
        }
        else {
            return null;
        }
    }
    /**
     * @return {?}
     */
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.use(this.fallbackLang);
            return this.use(this.activeLang);
        });
    }
    /**
     * Load the translation file or use provided data for the given language.
     *
     * \@memberof TranslateService
     * @param {?} lang Language name
     * @param {?=} data
     * @return {?} Final translation data merged with existing translations
     */
    use(lang, data) {
        return __awaiter(this, void 0, void 0, function* () {
            if (lang && data) {
                return this.setTranslation(lang, data);
            }
            /** @type {?} */
            let translation = this.data[lang];
            if (this.isNotSupported(lang)) {
                translation = this.data[this.fallbackLang];
            }
            if (translation && Object.keys(translation).length > 0) {
                return translation;
            }
            /** @type {?} */
            const fileName = `${lang || this.fallbackLang}.json`;
            /** @type {?} */
            const filePaths = [this.translationRoot, ...(this.translatePaths || [])];
            for (const path of filePaths) {
                /** @type {?} */
                const filePath = `${path}/${fileName}`;
                yield this.loadTranslation(lang, filePath);
            }
            return this.data[lang] || {};
        });
    }
    /**
     * @param {?} lang
     * @param {?} path
     * @return {?}
     */
    loadTranslation(lang, path) {
        return new Promise(resolve => {
            if (this.disableCache) {
                path += `?v=${Date.now()}`;
            }
            this.http.get(path).subscribe(json => {
                resolve(this.setTranslation(lang, json));
            }, () => {
                resolve(this.data[lang] || {});
            });
        });
    }
    /**
     * @param {?} lang
     * @return {?}
     */
    isNotSupported(lang) {
        return (lang !== this.fallbackLang &&
            lang !== this.activeLang &&
            (this.supportedLangs &&
                this.supportedLangs.length > 0 &&
                !this.supportedLangs.includes(lang)));
    }
    /**
     * @param {?} lang
     * @param {?} key
     * @return {?}
     */
    getValue(lang, key) {
        /** @type {?} */
        let data = this.data[lang];
        if (this.isNotSupported(lang)) {
            data = this.data[this.fallbackLang];
        }
        if (!data) {
            return key;
        }
        /** @type {?} */
        const keys = key.split('.');
        /** @type {?} */
        let propKey = '';
        do {
            propKey += keys.shift();
            /** @type {?} */
            const value = data[propKey];
            if (value !== undefined && (typeof value === 'object' || !keys.length)) {
                data = value;
                propKey = '';
            }
            else if (!keys.length) {
                data = key;
            }
            else {
                propKey += '.';
            }
        } while (keys.length);
        return data;
    }
    /**
     * @param {?} lang
     * @param {?} data
     * @return {?}
     */
    setTranslation(lang, data) {
        /** @type {?} */
        let finalResult = this.data[lang] || {};
        finalResult = this.merge(finalResult, data || {});
        this.data[lang] = finalResult;
        return finalResult;
    }
    /**
     * @param {...?} translations
     * @return {?}
     */
    merge(...translations) {
        /** @type {?} */
        const result = {};
        translations.forEach(translation => {
            Object.keys(translation).forEach(key => {
                if (key in result && Array.isArray(result[key])) {
                    result[key] = result[key].concat(translation[key]);
                }
                else if (key in result && typeof result[key] === 'object') {
                    result[key] = this.merge(result[key], translation[key]);
                }
                else {
                    result[key] = translation[key];
                }
            });
        });
        return result;
    }
    /**
     * @param {?} str
     * @param {?} params
     * @return {?}
     */
    format(str, params) {
        /** @type {?} */
        let result = str;
        if (params) {
            Object.keys(params).forEach(key => {
                /** @type {?} */
                const value = params[key];
                /** @type {?} */
                const template = new RegExp('{' + key + '}', 'gm');
                result = result.replace(template, value);
            });
        }
        if (this.debugMode) {
            result = `[${this.activeLang}] ${result}`;
        }
        return result;
    }
}
TranslateService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TranslateService.ctorParameters = () => [
    { type: HttpClient },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [TRANSLATE_SETTINGS,] }] }
];
/** @nocollapse */ TranslateService.ngInjectableDef = defineInjectable({ factory: function TranslateService_Factory() { return new TranslateService(inject(HttpClient), inject(TRANSLATE_SETTINGS, 8)); }, token: TranslateService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class TitleService {
    /**
     * @param {?} title
     * @param {?} translate
     */
    constructor(title, translate) {
        this.title = title;
        this.translate = translate;
        this._titleKey = '';
        this.changed = new Subject();
        translate.activeLangChanged.subscribe(() => this.onLanguageChanged());
    }
    /**
     * @return {?}
     */
    getTitle() {
        return this._titleKey;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    setTitle(key) {
        /** @type {?} */
        const previousValue = this._titleKey;
        /** @type {?} */
        const newValue = key || '';
        /** @type {?} */
        const changed = newValue !== previousValue;
        if (changed) {
            this._titleKey = newValue;
            /** @type {?} */
            const translated = this.translate.get(newValue);
            this.title.setTitle(translated);
            this.changed.next({
                previousValue: previousValue,
                currentValue: newValue
            });
        }
    }
    /**
     * @return {?}
     */
    onLanguageChanged() {
        if (this._titleKey) {
            /** @type {?} */
            const translated = this.translate.get(this._titleKey);
            this.title.setTitle(translated);
        }
    }
}
TitleService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TitleService.ctorParameters = () => [
    { type: Title },
    { type: TranslateService }
];
/** @nocollapse */ TitleService.ngInjectableDef = defineInjectable({ factory: function TitleService_Factory() { return new TitleService(inject(Title), inject(TranslateService)); }, token: TitleService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class TranslatePipe {
    /**
     * @param {?} translate
     */
    constructor(translate) {
        this.translate = translate;
    }
    /**
     * @param {?} key
     * @param {?=} params
     * @return {?}
     */
    transform(key, params) {
        return this.translate.get(key, params);
    }
}
TranslatePipe.decorators = [
    { type: Pipe, args: [{
                name: 'translate',
                pure: false
            },] }
];
/** @nocollapse */
TranslatePipe.ctorParameters = () => [
    { type: TranslateService }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class TranslateModule {
    /**
     * @param {?=} settings
     * @return {?}
     */
    static forRoot(settings) {
        return {
            ngModule: TranslateModule,
            providers: [
                { provide: TRANSLATE_SETTINGS, useValue: settings },
                TranslateService,
                TitleService
            ]
        };
    }
    /**
     * @return {?}
     */
    static forChild() {
        return {
            ngModule: TranslateModule
        };
    }
}
TranslateModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, HttpClientModule],
                declarations: [TranslatePipe],
                exports: [TranslatePipe]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { TRANSLATE_SETTINGS, TranslateService, TitleService, TranslatePipe, TranslateModule };
