// @ngstack/translate/esm2015/lib/translate.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { TranslateService, TRANSLATE_SETTINGS } from './translate.service';
import { TranslatePipe } from './translate.pipe';
import { TitleService } from './title.service';
export class TranslateModule {
    /**
     * @param {?=} settings
     * @return {?}
     */
    static forRoot(settings) {
        return {
            ngModule: TranslateModule,
            providers: [
                { provide: TRANSLATE_SETTINGS, useValue: settings },
                TranslateService,
                TitleService
            ]
        };
    }
    /**
     * @return {?}
     */
    static forChild() {
        return {
            ngModule: TranslateModule
        };
    }
}
TranslateModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, HttpClientModule],
                declarations: [TranslatePipe],
                exports: [TranslatePipe]
            },] }
];
