// @ngstack/translate/esm5/lib/title.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { TranslateService } from './translate.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
import * as i2 from "./translate.service";
var TitleService = /** @class */ (function () {
    function TitleService(title, translate) {
        var _this = this;
        this.title = title;
        this.translate = translate;
        this._titleKey = '';
        this.changed = new Subject();
        translate.activeLangChanged.subscribe(function () { return _this.onLanguageChanged(); });
    }
    /**
     * @return {?}
     */
    TitleService.prototype.getTitle = /**
     * @return {?}
     */
    function () {
        return this._titleKey;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    TitleService.prototype.setTitle = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var previousValue = this._titleKey;
        /** @type {?} */
        var newValue = key || '';
        /** @type {?} */
        var changed = newValue !== previousValue;
        if (changed) {
            this._titleKey = newValue;
            /** @type {?} */
            var translated = this.translate.get(newValue);
            this.title.setTitle(translated);
            this.changed.next({
                previousValue: previousValue,
                currentValue: newValue
            });
        }
    };
    /**
     * @return {?}
     */
    TitleService.prototype.onLanguageChanged = /**
     * @return {?}
     */
    function () {
        if (this._titleKey) {
            /** @type {?} */
            var translated = this.translate.get(this._titleKey);
            this.title.setTitle(translated);
        }
    };
    TitleService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TitleService.ctorParameters = function () { return [
        { type: Title },
        { type: TranslateService }
    ]; };
    /** @nocollapse */ TitleService.ngInjectableDef = i0.defineInjectable({ factory: function TitleService_Factory() { return new TitleService(i0.inject(i1.Title), i0.inject(i2.TranslateService)); }, token: TitleService, providedIn: "root" });
    return TitleService;
}());
export { TitleService };
if (false) {
    /** @type {?} */
    TitleService.prototype._titleKey;
    /** @type {?} */
    TitleService.prototype.changed;
    /** @type {?} */
    TitleService.prototype.title;
    /** @type {?} */
    TitleService.prototype.translate;
}
